### ISSUE TITLE
Documentation of backprocessing and forward processing in Substreams

### ISSUE THREAD

**Issue by maoueh** [streamingfast/substreams#68]
Start description of backprocessing and forward processing, I've written some content about those in Discord, would be great to bring them together in the documentation explaining the features and when they are activated exactly. Forward processing is dependent on production mode, as it will run only on this mode.

Not sure of exactly where to add data, your task will be to come up with a plan of where to integrate that as well as crawling Discord to generate the content and fill your gaps with knowledge transfer from someone on the team, maybe me.

**Comment by seanmooretechwriter** [comment]
I'll harvest the backprocessing threads and content from Discord first. Then clean it up. Then figure out a game plan for integrating it into the current docs.

**Comment by maoueh** [comment]
So, we have content that have been written now that we are getting closer to a new release of Substreams: https://github.com/streamingfast/substreams/blob/develop/docs/release-notes/change-log.md That should be the base for the explanation in the doc.

We also agreed we the dev team internally to **avoid** documenting explicitly the terms `back processing` and `forward processing`. We will talk about parallel execution (which is the process of chunking the chain in segment of N blocks and execute N workers over this range). 

The distinction in behavior will be on when you request the data relative to module's start block and if production vs dev mode is active or not.

Please read the material and we will find someone to help you with further questions you would have.

**Comment by seanmooretechwriter** [comment]
Two more:

In its current state, backprocessing has limitations. -- What are they?

Explain the features and when they are activated exactly. -- What are the features?

**Comment by maoueh** [comment]
So foremost, we are going to remove the term "backprocessing", which will be replace by "parallel execution"

> My question now is, what should backprocessing be referred to as? And what should the title of this page in the documentation be? Just simply "parallel execution?"

`Parallel Execution` is the term

> What is backprocessing?
Parallel execution is how your module's code is executed on multiple segments of the chain at the same time. In the server, we first define an execution schedule taken into account module's dependencies. The schedule actually a list of pair (`<module>`, `<range>`) where `<range>` contains N blocks, currently set on **our** server to 25K blocks (configurable). So for a single module `map_transfer`, if a use request from `0 - 75,000` then the execution plan would be: `[(map_transfer, 0 -> 24,999), (map_transfer, 25,000 -> 74,999), (map_transfer, 50,000 -> 74,999)]`.

Those 3 pairs will be executed all at the same time by the server, taking care of caching the output or the store. For `store`, a later step will combine back the store keys across multiple segment, leading to a unified and linear view of the store state.

Now for a full chain segment, assuming Ethereum Mainnet with ~16,000,000 blocks, it means 640 segments of 25K blocks. Of course, the server has a limit on the amount of concurrency it's willing to accept. Theoretically, 640 concurrent workers could be spanned. Practically, it depends on the service provider. We set our concurrency at 15 on production endpoint for fair usage of resources while the service is free.

> Why is backprocessing important?

This is what make the Substreams super fast. To give an idea, with 120 concurrent workers, we were able to full process and store Ethereum Mainnet for all ERC20/ERC721/ERC1155 transfers in ~45m. 

> What is backprocessing used for?

Used to pre-compute module's execution. Since modules are executed in parallel, it means data is ready rapidly for user's consumption.

Parallel execution happens:
- When requested module's start block is in the past relative to requested start block (module starts at 12,000,000, user request data at block 15,000,000) and both mode (development & production)
- When in production mode, parallel execution is performed for the full range.

> What problems does backprocessing solve?

Solves slow single linear execution of a module. Instead of running a module in a linear fashion, one block after the other without leverage full computing power, we execute N workers over different segment of the chain. It means we are able to push data back to the user N times faster than if we had 1 worker.

> How does backprocessing work (high-level)?
> How does backprocessing work (detailed, deep dive)?

Explained in first paragraph

> What is not addressed through backprocessing?

I'm not sure what it means
